### PR TITLE
start Week04

### PR DESCRIPTION
**Pull request by armando-nava-it**

### CODE REVIEW

File: src/Week04/HW04.hs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+x :: Num a => Poly a
+x = P [0, 1]
+
+-- Exercise 2 ----------------------------------------
+
+instance (Eq a, Num a) => Eq (Poly a) where
+    (P xs) == (P ys) = trim xs == trim ys
+      where
+        trim = dropWhile (== 0) . reverse
```
Comment (jhrcek) [inline src/Week04/HW04.hs]: There's also [dropWhileEnd](https://hackage.haskell.org/package/base-4.21.0.0/docs/Data-List.html#v:dropWhileEnd), but this implementation is also fine.

File: src/Week04/HW04.hs [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+-- Exercise 3 -----------------------------------------
+
+instance (Eq a, Num a, Show a) => Show (Poly a) where
+    show (P coeffs) =
+        let terms = reverse $ zip [0 :: Int ..] coeffs
+            shownTerms = map showTerm $ filter (\(_, c) -> c /= 0) terms
+         in if null shownTerms
+                then "0"
+                else unwords $ insertPluses shownTerms
```
Comment (jhrcek) [inline src/Week04/HW04.hs]: You can simplify this by using [intercalate](https://hackage.haskell.org/package/base-4.21.0.0/docs/Data-List.html#v:intercalate)

File: src/Week04/HW04.hs [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    nderiv n f
+        | n <= 0 = f
+        | otherwise = deriv (nderiv (n - 1) f)
+
+-- Exercise 9 -----------------------------------------
+
+instance (Enum a, Num a) => Differentiable (Poly a) where
+    deriv :: Poly a -> Poly a
+    deriv (P coeffs) =
+        let derivCoeffs = zipWith (*) (map fromIntegral [1 :: Int ..]) (tail coeffs)
```
Comment (jhrcek) [inline src/Week04/HW04.hs]: tail is a partial function and should be generally avoided (newer GHC version even give you a warning whenever you use it). Can you fix this implementation so it also works  for the edge case of an empty list (which should be equivalent to polynomial `f(x) = 0`)?

File: test/HW04Tests.hs [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+import Test.Hspec
+
+import Week04.HW04
+
+allTests :: Spec
+allTests = describe "Week04" $ do
+    describe "\nPoly Eq Instance" $ do
+        it "Identical Poly " $ do
+            P ([1, 2] :: [Int]) `shouldBe` P ([1, 2, 0] :: [Int])
```
Comment (jhrcek) [inline test/HW04Tests.hs]: Adding type annotations everywhere seems like an overkill.
You can take advantage of type inference and only add type annotation to the first element of one of the list (and the rest is inferred by the compiler). It's less typing and fewer places to change if you decide to change the representation.
```suggestion
            P [1 :: Int, 2] `shouldBe` P [1, 2, 0]
```

File: test/HW04Tests.hs [hunk 5]
```diff
@@ -0,0 +1,51 @@
+{-# LANGUAGE NondecreasingIndentation #-}
```
Comment (jhrcek) [inline test/HW04Tests.hs]: This shouldn't be needed if you format things properly (and it should be easy to forma things properly if you use formatter like fourmolu :slightly_smiling_face: )
```suggestion
```

File: src/Week04/HW04.hs [hunk 6]
```diff
@@ -0,0 +1,98 @@
+{-# LANGUAGE InstanceSigs #-}
+{-# OPTIONS_GHC -Wall #-}
```
Comment (jhrcek) [inline src/Week04/HW04.hs]: This is redundant, as we have -Wall enabled for all files in the project (see -Wall occurrences in the cabal file).
```suggestion
```

**Review by jhrcek — COMMENTED** [review]
Looks good, just bunch of minor comments.

**Review by jhrcek — APPROVED** [review]
Thanks, Looks good to me.
